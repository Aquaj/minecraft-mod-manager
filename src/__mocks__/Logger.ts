import { vi } from 'vitest';

export const Logger = vi.importMock('../lib/Logger.js');
