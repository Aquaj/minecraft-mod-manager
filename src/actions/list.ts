import { readConfigFile } from '../lib/config.js';
import chalk from 'chalk';
import { DefaultOptions } from '../mmu.js';

export interface ListOptions extends DefaultOptions {}

export const list = async (options: ListOptions) => {
  const config = await readConfigFile(options.config);

  console.log((chalk.green('Configured mods')));

  config.mods.forEach((mod) => {
    console.log(` ${mod.name}`);
    if (mod.installed) {
      console.log('   ', chalk.green('\u2705'), 'Installed');
    } else {
      console.log('   ', chalk.green('\u274c'), 'Not installed');
    }

  });
};
