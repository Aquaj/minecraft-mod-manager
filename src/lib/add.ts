import path from 'path';
import { DefaultOptions } from '../index.js';
import { fetchModDetails } from '../repositories/index.js';
import { ModConfig, Platform } from './modlist.types.js';
import { readConfigFile, writeConfigFile } from './config.js';
import { downloadFile } from './downloader.js';

export interface AddOptions extends DefaultOptions {
}

export const add = async (platform: Platform, id: string, options: AddOptions) => {
  const configuration = await readConfigFile(options.config);

  const moddata = await fetchModDetails(platform, id, configuration.defaultAllowedReleaseTypes, configuration.gameVersion, configuration.loader);

  await downloadFile(moddata.downloadUrl, path.resolve(configuration.modsFolder, moddata.fileName));

  if (configuration.mods.find((mod: ModConfig) => (mod.id === id && mod.type === platform))) {
    return;
  }

  configuration.mods.push({
    type: platform,
    id: id,
    name: moddata.name,
    installed: {
      fileName: moddata.fileName,
      releasedOn: moddata.releaseDate,
      hash: moddata.hash
    },
    allowedReleaseTypes: configuration.defaultAllowedReleaseTypes
  });

  await writeConfigFile(configuration, options.config);

};
