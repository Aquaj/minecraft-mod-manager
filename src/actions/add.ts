import path from 'path';
import { fetchModDetails } from '../repositories/index.js';
import { Mod, Platform } from '../lib/modlist.types.js';
import {
  ensureConfiguration,
  readLockFile,
  writeConfigFile,
  writeLockFile
} from '../lib/config.js';
import { downloadFile } from '../lib/downloader.js';
import { DefaultOptions, stop } from '../mmm.js';
import { UnknownPlatformException } from '../errors/UnknownPlatformException.js';
import inquirer from 'inquirer';
import chalk from 'chalk';
import { CouldNotFindModException } from '../errors/CouldNotFindModException.js';
import { NoRemoteFileFound } from '../errors/NoRemoteFileFound.js';
import { Logger } from '../lib/Logger.js';
import { modNotFound } from '../interactions/modNotFound.js';
import { noRemoteFileFound } from '../interactions/noRemoteFileFound.js';
import { DownloadFailedException } from '../errors/DownloadFailedException.js';

const handleUnknownPlatformException = async (error: UnknownPlatformException, id: string, options: DefaultOptions, logger: Logger) => {
  const platformUsed = error.platform;
  const platformList = Object.values(Platform);

  if (options.quiet === true) {
    logger.error(`Unknown platform "${chalk.whiteBright(platformUsed)}". Please use one of the following: ${chalk.whiteBright(platformList.join(', '))}`);
  }

  const answers = await inquirer.prompt([
    {
      type: 'list',
      name: 'platform',
      default: false,
      choices: [...platformList, 'cancel'],
      message: chalk.redBright(`The platform you entered (${chalk.whiteBright(platformUsed)}) is not a valid platform.\n`)
        + chalk.whiteBright('Would you like to retry with a valid one?')
    }
  ]);

  if (answers.platform === 'cancel') {
    stop();
  }

  // eslint-disable-next-line no-use-before-define
  await add(answers.platform, id, options, logger);

};

export const add = async (platform: Platform, id: string, options: DefaultOptions, logger: Logger) => {

  const configuration = await ensureConfiguration(options.config, logger, options.quiet);
  const modConfig = configuration.mods.find((mod: Mod) => (mod.id === id && mod.type === platform));

  if (modConfig) {
    logger.debug(`Mod ${id} for ${platform} already exists in the configuration`);
    return;
  }

  try {
    const modData = await fetchModDetails(
      platform,
      id,
      configuration.defaultAllowedReleaseTypes,
      configuration.gameVersion,
      configuration.loader,
      configuration.allowVersionFallback);

    await downloadFile(modData.downloadUrl, path.resolve(configuration.modsFolder, modData.fileName));

    const installations = await readLockFile(options, logger);

    configuration.mods.push({
      type: platform,
      id: id,
      name: modData.name
    });

    installations.push({
      name: modData.name,
      type: platform,
      id: id,
      fileName: modData.fileName,
      releasedOn: modData.releaseDate,
      hash: modData.hash,
      downloadUrl: modData.downloadUrl
    });

    await writeConfigFile(configuration, options, logger);
    await writeLockFile(installations, options, logger);

    logger.log(`${chalk.green('\u2705')} Added ${modData.name} (${id}) for ${platform}`);

  } catch (error) {
    if (error instanceof DownloadFailedException) {
      logger.error(error.message, 1);
    }
    if (error instanceof UnknownPlatformException) {
      await handleUnknownPlatformException(error, id, options, logger);
      return;
    }

    if (error instanceof CouldNotFindModException) {
      const { id: newId, platform: newPlatform } = await modNotFound(id, platform, logger, options);
      await add(newPlatform, newId, options, logger);
      return;
    }

    if (error instanceof NoRemoteFileFound) {
      const {
        id: newId,
        platform: newPlatform
      } = await noRemoteFileFound(id, platform, configuration, logger, options);
      await add(newPlatform, newId, options, logger);
      return;
    }

    logger.error((error as Error).message, 2);
  }

};
