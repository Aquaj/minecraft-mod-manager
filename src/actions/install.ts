import {
  fileExists,
  readConfigFile,
  readLockFile,
  writeConfigFile,
  writeLockFile
} from '../lib/config.js';
import path from 'path';
import { fetchModDetails } from '../repositories/index.js';
import { downloadFile } from '../lib/downloader.js';
import { Mod, RemoteModDetails } from '../lib/modlist.types.js';
import { getHash } from '../lib/hash.js';
import { DefaultOptions } from '../mmm.js';
import { updateMod } from '../lib/updater.js';
import { Logger } from '../lib/Logger.js';
import { ConfigFileNotFoundException } from '../errors/ConfigFileNotFoundException.js';
import { ErrorTexts } from '../errors/ErrorTexts.js';
import { getInstallation, hasInstallation } from '../lib/configurationHelper.js';

const getMod = async (moddata: RemoteModDetails, modsFolder: string) => {
  await downloadFile(moddata.downloadUrl, path.resolve(modsFolder, moddata.fileName));
  return {
    fileName: moddata.fileName,
    releasedOn: moddata.releaseDate,
    hash: moddata.hash,
    downloadUrl: moddata.downloadUrl
  };
};

export const install = async (options: DefaultOptions, logger: Logger) => {
  try {
    const configuration = await readConfigFile(options.config);
    const installations = await readLockFile(options.config);

    const installedMods = installations;
    const mods = configuration.mods;

    const processMod = async (mod: Mod, index: number) => {

      logger.debug(`Checking ${mod.name} for ${mod.type}`);

      if (hasInstallation(mod, installations)) {
        const installedModIndex = getInstallation(mod, installedMods);

        const modPath = path.resolve(configuration.modsFolder, installedMods[installedModIndex].fileName);

        if (!await fileExists(modPath)) {
          logger.log(`${mod.name} doesn't exist, downloading from ${installedMods[installedModIndex].type}`);
          await downloadFile(installedMods[installedModIndex].downloadUrl, modPath);
          // TODO handle the download failing
          return;
        }

        const installedHash = await getHash(modPath);
        if (installedMods[installedModIndex].hash !== installedHash) {
          logger.log(`${mod.name} has hash mismatch, downloading from source`);
          await updateMod(installedMods[installedModIndex], modPath, configuration.modsFolder);
          // TODO handle the download failing
          return;
        }
        return;
      }

      const modData = await fetchModDetails(
        mod.type,
        mod.id,
        mod.allowedReleaseTypes || configuration.defaultAllowedReleaseTypes,
        configuration.gameVersion,
        configuration.loader,
        configuration.allowVersionFallback
      );
      // TODO Handle the fetch failing

      mods[index].name = modData.name;

      // no installation exists
      logger.log(`${mod.name} doesn't exist, downloading from ${mod.type}`);
      const dlData = await getMod(modData, configuration.modsFolder);
      // TODO handle the download failing
      installedMods.push({
        name: modData.name,
        type: mod.type,
        id: mod.id,
        fileName: dlData.fileName,
        releasedOn: dlData.releasedOn,
        hash: dlData.hash,
        downloadUrl: dlData.downloadUrl
      });
      return;

    };

    const promises = mods.map(processMod);

    await Promise.all(promises);

    await writeLockFile(installedMods, options.config);
    await writeConfigFile(configuration, options.config);
  } catch (error) {
    if (error instanceof ConfigFileNotFoundException) {
      logger.error(ErrorTexts.configNotFound);
    }

    logger.error((error as Error).message, 2);
  }
};
