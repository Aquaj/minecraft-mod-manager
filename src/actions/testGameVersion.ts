import chalk from 'chalk';
import { IncorrectMinecraftVersionException } from '../errors/IncorrectMinecraftVersionException.js';
import { RedundantVersionException } from '../errors/RedundantVersionException.js';
import { Logger } from '../lib/Logger.js';
import { verifyUpgradeIsPossible, VerifyUpgradeOptions } from '../lib/verifyUpgrade.js';
import { EXIT_CODE } from '../mmm.js';

export const testGameVersion = async (gameVersion: string, options: VerifyUpgradeOptions, logger: Logger) => {
  try {
    const { canUpgrade, version, modsInError } = await verifyUpgradeIsPossible(gameVersion, options, logger);
    if (!canUpgrade) {
      logger.log(`Some mods are missing support for ${version}.`);
      modsInError.forEach((mod) => {
        logger.log(`${chalk.red('\u274c')} ${mod.name?.trim()}`);
      });
      logger.error(`You cannot upgrade to ${version} just yet.`, 1);
    } else {
      logger.log(chalk.green(`All mods have support for ${version}. You can safely upgrade.`));
    }
  } catch (e) {
    if (e instanceof IncorrectMinecraftVersionException) {
      logger.error(e.message, EXIT_CODE.GENERAL_ERROR);
    }
    if (e instanceof RedundantVersionException) {
      logger.error(e.message, EXIT_CODE.SUPPLEMENTARY_ERROR);
    }
  }
};
