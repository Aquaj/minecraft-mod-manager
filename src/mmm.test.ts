import { chance } from 'jest-chance';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { add } from './actions/add.js';
import { install } from './actions/install.js';
import { list } from './actions/list.js';
import { testGameVersion } from './actions/testGameVersion.js';
import { update } from './actions/update.js';
import { initializeConfig } from './interactions/initializeConfig.js';
import { Logger } from './lib/Logger.js';
import { Platform } from './lib/modlist.types.js';

vi.mock('./lib/Logger.js');
vi.mock('./actions/add.js');
vi.mock('./actions/list.js');
vi.mock('./actions/install.js');
vi.mock('./actions/update.js');
vi.mock('./interactions/initializeConfig.js');
vi.mock('./actions/testGameVersion.js');

describe('The main CLI configuration', () => {
  let logger: Logger;
  beforeEach(() => {
    vi.resetModules();
    vi.spyOn(process, 'cwd').mockReturnValue('/path/to/minecraft/installation');
    logger = new Logger({} as never);
  });

  afterEach(() => {
    vi.resetAllMocks();
    vi.resetModules();
  });

  it('is set up correctly', async () => {
    const { program } = await import('./mmm.js');
    expect(program).toMatchSnapshot();
  });

  it('has add hooked up to the correct function', async () => {
    const { program } = await import('./mmm.js');
    vi.mocked(add).mockResolvedValueOnce();
    await program.parse([
      '',
      '',
      chance.pickone(['add', 'a']),
      chance.pickone(Object.values(Platform)),
      chance.word()]);
    expect(vi.mocked(add)).toHaveBeenCalledOnce();
  });

  it('has list hooked up to the correct function', async () => {
    const { program } = await import('./mmm.js');

    vi.mocked(list).mockResolvedValueOnce();
    await program.parse([
      '',
      '',
      chance.pickone(['list', 'ls'])
    ]);
    expect(vi.mocked(list)).toHaveBeenCalledOnce();
  });

  it('has install hooked up to the correct function', async () => {
    const { program } = await import('./mmm.js');

    vi.mocked(install).mockResolvedValueOnce();
    await program.parse([
      '',
      '',
      chance.pickone(['install', 'i'])
    ]);
    expect(vi.mocked(install)).toHaveBeenCalledOnce();
  });

  it('has update hooked up to the correct function', async () => {
    const { program } = await import('./mmm.js');

    vi.mocked(update).mockResolvedValueOnce();
    await program.parse([
      '',
      '',
      chance.pickone(['update', 'u'])
    ]);
    expect(vi.mocked(update)).toHaveBeenCalledOnce();
  });

  it('has initialize hooked up to the correct function', async () => {
    const { program } = await import('./mmm.js');

    vi.mocked(initializeConfig).mockResolvedValueOnce(expect.anything());
    await program.parse([
      '',
      '',
      chance.pickone(['init'])
    ]);
    expect(vi.mocked(initializeConfig)).toHaveBeenCalledOnce();
  });

  it('has the test hooked up to the correct function', async () => {
    const { program } = await import('./mmm.js');
    vi.mocked(testGameVersion).mockResolvedValueOnce(expect.anything());
    await program.parse([
      '',
      '',
      chance.pickone(['test', 't'])
    ]);
    expect(testGameVersion).toHaveBeenCalledOnce();
  });

  it('sets the logger to quiet when the quiet option is supplied', async () => {
    const { program } = await import('./mmm.js');
    await program.parse([
      '',
      '',
      chance.pickone(['-q', '--quiet']),
      chance.pickone(['init'])
    ]);
    expect(logger.flagQuiet).toHaveBeenCalledOnce();
  });

  it('sets the logger to debug when the debug option is supplied', async () => {
    const { program } = await import('./mmm.js');
    await program.parse([
      '',
      '',
      chance.pickone(['-d', '--debug']),
      chance.pickone(['init'])
    ]);
    expect(logger.flagDebug).toHaveBeenCalledOnce();
  });

  it('can stop the execution', async () => {
    vi.spyOn(process, 'exit').mockImplementation(() => {
      throw new Error('process.exit');
    });
    const { stop } = await import('./mmm.js');
    expect(stop).toThrow(new Error('process.exit'));
  });
});
