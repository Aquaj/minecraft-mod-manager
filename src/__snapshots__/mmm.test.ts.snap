// Vitest Snapshot v1

exports[`The main CLI configuration > is set up correctly 1`] = `
Command {
  "_actionHandler": null,
  "_addImplicitHelpCommand": undefined,
  "_aliases": [],
  "_allowExcessArguments": true,
  "_allowUnknownOption": false,
  "_args": [],
  "_argsDescription": undefined,
  "_combineFlagAndOptionalValue": true,
  "_defaultCommandName": null,
  "_description": "Manages mods from Modrinth and Curseforge",
  "_enablePositionalOptions": false,
  "_events": {
    "afterHelp": [Function],
    "option:config": [Function],
    "option:debug": [
      [Function],
      [Function],
    ],
    "option:quiet": [
      [Function],
      [Function],
    ],
    "option:version": [Function],
  },
  "_eventsCount": 5,
  "_executableDir": null,
  "_executableFile": null,
  "_executableHandler": false,
  "_exitCallback": null,
  "_hasHelpOption": true,
  "_helpCommandDescription": "display help for command",
  "_helpCommandName": "help",
  "_helpCommandnameAndArgs": "help [command]",
  "_helpConfiguration": {},
  "_helpDescription": "display help for command",
  "_helpFlags": "-h, --help",
  "_helpLongFlag": "--help",
  "_helpShortFlag": "-h",
  "_hidden": false,
  "_lifeCycleHooks": {},
  "_maxListeners": undefined,
  "_name": "Minecraft Mod Manager",
  "_optionValueSources": {
    "config": "default",
    "debug": "default",
    "quiet": "default",
  },
  "_optionValues": {
    "config": "./modlist.json",
    "debug": false,
    "quiet": false,
  },
  "_outputConfiguration": {
    "getErrHelpWidth": [Function],
    "getOutHelpWidth": [Function],
    "outputError": [Function],
    "writeErr": [Function],
    "writeOut": [Function],
  },
  "_passThroughOptions": false,
  "_scriptPath": null,
  "_showHelpAfterError": false,
  "_showSuggestionAfterError": true,
  "_storeOptionsAsProperties": false,
  "_summary": "",
  "_version": "DEV",
  "_versionOptionName": "version",
  "args": [],
  "commands": [
    Command {
      "_actionHandler": [Function],
      "_addImplicitHelpCommand": undefined,
      "_aliases": [
        "l",
        "ls",
      ],
      "_allowExcessArguments": true,
      "_allowUnknownOption": false,
      "_args": [],
      "_argsDescription": undefined,
      "_combineFlagAndOptionalValue": true,
      "_defaultCommandName": null,
      "_description": "",
      "_enablePositionalOptions": false,
      "_events": {},
      "_eventsCount": 0,
      "_executableDir": null,
      "_executableFile": null,
      "_executableHandler": false,
      "_exitCallback": null,
      "_hasHelpOption": true,
      "_helpCommandDescription": "display help for command",
      "_helpCommandName": "help",
      "_helpCommandnameAndArgs": "help [command]",
      "_helpConfiguration": {},
      "_helpDescription": "display help for command",
      "_helpFlags": "-h, --help",
      "_helpLongFlag": "--help",
      "_helpShortFlag": "-h",
      "_hidden": false,
      "_lifeCycleHooks": {},
      "_maxListeners": undefined,
      "_name": "list",
      "_optionValueSources": {},
      "_optionValues": {},
      "_outputConfiguration": {
        "getErrHelpWidth": [Function],
        "getOutHelpWidth": [Function],
        "outputError": [Function],
        "writeErr": [Function],
        "writeOut": [Function],
      },
      "_passThroughOptions": false,
      "_scriptPath": null,
      "_showHelpAfterError": false,
      "_showSuggestionAfterError": true,
      "_storeOptionsAsProperties": false,
      "_summary": "",
      "args": [],
      "commands": [],
      "options": [],
      "parent": [Circular],
      "processedArgs": [],
      "rawArgs": [],
      Symbol(kCapture): false,
    },
    Command {
      "_actionHandler": [Function],
      "_addImplicitHelpCommand": undefined,
      "_aliases": [
        "i",
      ],
      "_allowExcessArguments": true,
      "_allowUnknownOption": false,
      "_args": [],
      "_argsDescription": undefined,
      "_combineFlagAndOptionalValue": true,
      "_defaultCommandName": null,
      "_description": "",
      "_enablePositionalOptions": false,
      "_events": {},
      "_eventsCount": 0,
      "_executableDir": null,
      "_executableFile": null,
      "_executableHandler": false,
      "_exitCallback": null,
      "_hasHelpOption": true,
      "_helpCommandDescription": "display help for command",
      "_helpCommandName": "help",
      "_helpCommandnameAndArgs": "help [command]",
      "_helpConfiguration": {},
      "_helpDescription": "display help for command",
      "_helpFlags": "-h, --help",
      "_helpLongFlag": "--help",
      "_helpShortFlag": "-h",
      "_hidden": false,
      "_lifeCycleHooks": {},
      "_maxListeners": undefined,
      "_name": "install",
      "_optionValueSources": {},
      "_optionValues": {},
      "_outputConfiguration": {
        "getErrHelpWidth": [Function],
        "getOutHelpWidth": [Function],
        "outputError": [Function],
        "writeErr": [Function],
        "writeOut": [Function],
      },
      "_passThroughOptions": false,
      "_scriptPath": null,
      "_showHelpAfterError": false,
      "_showSuggestionAfterError": true,
      "_storeOptionsAsProperties": false,
      "_summary": "",
      "args": [],
      "commands": [],
      "options": [],
      "parent": [Circular],
      "processedArgs": [],
      "rawArgs": [],
      Symbol(kCapture): false,
    },
    Command {
      "_actionHandler": [Function],
      "_addImplicitHelpCommand": undefined,
      "_aliases": [
        "u",
      ],
      "_allowExcessArguments": true,
      "_allowUnknownOption": false,
      "_args": [],
      "_argsDescription": undefined,
      "_combineFlagAndOptionalValue": true,
      "_defaultCommandName": null,
      "_description": "",
      "_enablePositionalOptions": false,
      "_events": {},
      "_eventsCount": 0,
      "_executableDir": null,
      "_executableFile": null,
      "_executableHandler": false,
      "_exitCallback": null,
      "_hasHelpOption": true,
      "_helpCommandDescription": "display help for command",
      "_helpCommandName": "help",
      "_helpCommandnameAndArgs": "help [command]",
      "_helpConfiguration": {},
      "_helpDescription": "display help for command",
      "_helpFlags": "-h, --help",
      "_helpLongFlag": "--help",
      "_helpShortFlag": "-h",
      "_hidden": false,
      "_lifeCycleHooks": {},
      "_maxListeners": undefined,
      "_name": "update",
      "_optionValueSources": {},
      "_optionValues": {},
      "_outputConfiguration": {
        "getErrHelpWidth": [Function],
        "getOutHelpWidth": [Function],
        "outputError": [Function],
        "writeErr": [Function],
        "writeOut": [Function],
      },
      "_passThroughOptions": false,
      "_scriptPath": null,
      "_showHelpAfterError": false,
      "_showSuggestionAfterError": true,
      "_storeOptionsAsProperties": false,
      "_summary": "",
      "args": [],
      "commands": [],
      "options": [],
      "parent": [Circular],
      "processedArgs": [],
      "rawArgs": [],
      Symbol(kCapture): false,
    },
    Command {
      "_actionHandler": [Function],
      "_addImplicitHelpCommand": undefined,
      "_aliases": [
        "a",
      ],
      "_allowExcessArguments": true,
      "_allowUnknownOption": false,
      "_args": [
        Argument {
          "_name": "type",
          "argChoices": undefined,
          "defaultValue": undefined,
          "defaultValueDescription": undefined,
          "description": "curseforge or modrinth",
          "parseArg": undefined,
          "required": true,
          "variadic": false,
        },
        Argument {
          "_name": "id",
          "argChoices": undefined,
          "defaultValue": undefined,
          "defaultValueDescription": undefined,
          "description": "Curseforge or Modrinth Project Id",
          "parseArg": undefined,
          "required": true,
          "variadic": false,
        },
      ],
      "_argsDescription": undefined,
      "_combineFlagAndOptionalValue": true,
      "_defaultCommandName": null,
      "_description": "",
      "_enablePositionalOptions": false,
      "_events": {},
      "_eventsCount": 0,
      "_executableDir": null,
      "_executableFile": null,
      "_executableHandler": false,
      "_exitCallback": null,
      "_hasHelpOption": true,
      "_helpCommandDescription": "display help for command",
      "_helpCommandName": "help",
      "_helpCommandnameAndArgs": "help [command]",
      "_helpConfiguration": {},
      "_helpDescription": "display help for command",
      "_helpFlags": "-h, --help",
      "_helpLongFlag": "--help",
      "_helpShortFlag": "-h",
      "_hidden": false,
      "_lifeCycleHooks": {},
      "_maxListeners": undefined,
      "_name": "add",
      "_optionValueSources": {},
      "_optionValues": {},
      "_outputConfiguration": {
        "getErrHelpWidth": [Function],
        "getOutHelpWidth": [Function],
        "outputError": [Function],
        "writeErr": [Function],
        "writeOut": [Function],
      },
      "_passThroughOptions": false,
      "_scriptPath": null,
      "_showHelpAfterError": false,
      "_showSuggestionAfterError": true,
      "_storeOptionsAsProperties": false,
      "_summary": "",
      "args": [],
      "commands": [],
      "options": [],
      "parent": [Circular],
      "processedArgs": [],
      "rawArgs": [],
      Symbol(kCapture): false,
    },
    Command {
      "_actionHandler": [Function],
      "_addImplicitHelpCommand": undefined,
      "_aliases": [],
      "_allowExcessArguments": true,
      "_allowUnknownOption": false,
      "_args": [],
      "_argsDescription": undefined,
      "_combineFlagAndOptionalValue": true,
      "_defaultCommandName": null,
      "_description": "",
      "_enablePositionalOptions": false,
      "_events": {
        "option:allow-version-fallback": [Function],
        "option:default-allowed-release-types": [Function],
        "option:game-version": [Function],
        "option:loader": [Function],
        "option:mods-folder": [Function],
      },
      "_eventsCount": 5,
      "_executableDir": null,
      "_executableFile": null,
      "_executableHandler": false,
      "_exitCallback": null,
      "_hasHelpOption": true,
      "_helpCommandDescription": "display help for command",
      "_helpCommandName": "help",
      "_helpCommandnameAndArgs": "help [command]",
      "_helpConfiguration": {},
      "_helpDescription": "display help for command",
      "_helpFlags": "-h, --help",
      "_helpLongFlag": "--help",
      "_helpShortFlag": "-h",
      "_hidden": false,
      "_lifeCycleHooks": {},
      "_maxListeners": undefined,
      "_name": "init",
      "_optionValueSources": {},
      "_optionValues": {},
      "_outputConfiguration": {
        "getErrHelpWidth": [Function],
        "getOutHelpWidth": [Function],
        "outputError": [Function],
        "writeErr": [Function],
        "writeOut": [Function],
      },
      "_passThroughOptions": false,
      "_scriptPath": null,
      "_showHelpAfterError": false,
      "_showSuggestionAfterError": true,
      "_storeOptionsAsProperties": false,
      "_summary": "",
      "args": [],
      "commands": [],
      "options": [
        Option {
          "argChoices": undefined,
          "conflictsWith": [],
          "defaultValue": undefined,
          "defaultValueDescription": undefined,
          "description": "Which loader would you like to use? curseforge, modrinth",
          "envVar": undefined,
          "flags": "-l, --loader <loader>",
          "hidden": false,
          "implied": undefined,
          "long": "--loader",
          "mandatory": false,
          "negate": false,
          "optional": false,
          "parseArg": undefined,
          "presetArg": undefined,
          "required": true,
          "short": "-l",
          "variadic": false,
        },
        Option {
          "argChoices": undefined,
          "conflictsWith": [],
          "defaultValue": undefined,
          "defaultValueDescription": undefined,
          "description": "What exact Minecraft version are you using? (eg: 1.18.2, 1.19, 1.19.1)",
          "envVar": undefined,
          "flags": "-g, --game-version <gameVersion>",
          "hidden": false,
          "implied": undefined,
          "long": "--game-version",
          "mandatory": false,
          "negate": false,
          "optional": false,
          "parseArg": undefined,
          "presetArg": undefined,
          "required": true,
          "short": "-g",
          "variadic": false,
        },
        Option {
          "argChoices": undefined,
          "conflictsWith": [],
          "defaultValue": undefined,
          "defaultValueDescription": undefined,
          "description": "Should we try to download mods for previous Minecraft versions if they do not exists for your Minecraft Version?",
          "envVar": undefined,
          "flags": "-f, --allow-version-fallback",
          "hidden": false,
          "implied": undefined,
          "long": "--allow-version-fallback",
          "mandatory": false,
          "negate": false,
          "optional": false,
          "parseArg": undefined,
          "presetArg": undefined,
          "required": false,
          "short": "-f",
          "variadic": false,
        },
        Option {
          "argChoices": undefined,
          "conflictsWith": [],
          "defaultValue": undefined,
          "defaultValueDescription": undefined,
          "description": "Which types of releases would you like to consider to download? alpha, beta, release - comma separated list",
          "envVar": undefined,
          "flags": "-r, --default-allowed-release-types <defaultAllowedReleaseTypes>",
          "hidden": false,
          "implied": undefined,
          "long": "--default-allowed-release-types",
          "mandatory": false,
          "negate": false,
          "optional": false,
          "parseArg": undefined,
          "presetArg": undefined,
          "required": true,
          "short": "-r",
          "variadic": false,
        },
        Option {
          "argChoices": undefined,
          "conflictsWith": [],
          "defaultValue": undefined,
          "defaultValueDescription": undefined,
          "description": "where is your mods folder? (full or relative path from /path/to/minecraft/installation)",
          "envVar": undefined,
          "flags": "-m, --mods-folder <modsFolder>",
          "hidden": false,
          "implied": undefined,
          "long": "--mods-folder",
          "mandatory": false,
          "negate": false,
          "optional": false,
          "parseArg": undefined,
          "presetArg": undefined,
          "required": true,
          "short": "-m",
          "variadic": false,
        },
      ],
      "parent": [Circular],
      "processedArgs": [],
      "rawArgs": [],
      Symbol(kCapture): false,
    },
    Command {
      "_actionHandler": [Function],
      "_addImplicitHelpCommand": undefined,
      "_aliases": [
        "t",
      ],
      "_allowExcessArguments": true,
      "_allowUnknownOption": false,
      "_args": [
        Argument {
          "_name": "game_version",
          "argChoices": undefined,
          "defaultValue": "latest",
          "defaultValueDescription": undefined,
          "description": "The Minecraft version to test",
          "parseArg": undefined,
          "required": false,
          "variadic": false,
        },
      ],
      "_argsDescription": undefined,
      "_combineFlagAndOptionalValue": true,
      "_defaultCommandName": null,
      "_description": "",
      "_enablePositionalOptions": false,
      "_events": {},
      "_eventsCount": 0,
      "_executableDir": null,
      "_executableFile": null,
      "_executableHandler": false,
      "_exitCallback": null,
      "_hasHelpOption": true,
      "_helpCommandDescription": "display help for command",
      "_helpCommandName": "help",
      "_helpCommandnameAndArgs": "help [command]",
      "_helpConfiguration": {},
      "_helpDescription": "display help for command",
      "_helpFlags": "-h, --help",
      "_helpLongFlag": "--help",
      "_helpShortFlag": "-h",
      "_hidden": false,
      "_lifeCycleHooks": {},
      "_maxListeners": undefined,
      "_name": "test",
      "_optionValueSources": {},
      "_optionValues": {},
      "_outputConfiguration": {
        "getErrHelpWidth": [Function],
        "getOutHelpWidth": [Function],
        "outputError": [Function],
        "writeErr": [Function],
        "writeOut": [Function],
      },
      "_passThroughOptions": false,
      "_scriptPath": null,
      "_showHelpAfterError": false,
      "_showSuggestionAfterError": true,
      "_storeOptionsAsProperties": false,
      "_summary": "",
      "args": [],
      "commands": [],
      "options": [],
      "parent": [Circular],
      "processedArgs": [],
      "rawArgs": [],
      Symbol(kCapture): false,
    },
  ],
  "options": [
    Option {
      "argChoices": undefined,
      "conflictsWith": [],
      "defaultValue": undefined,
      "defaultValueDescription": undefined,
      "description": "output the version number",
      "envVar": undefined,
      "flags": "-V, --version",
      "hidden": false,
      "implied": undefined,
      "long": "--version",
      "mandatory": false,
      "negate": false,
      "optional": false,
      "parseArg": undefined,
      "presetArg": undefined,
      "required": false,
      "short": "-V",
      "variadic": false,
    },
    Option {
      "argChoices": undefined,
      "conflictsWith": [],
      "defaultValue": "./modlist.json",
      "defaultValueDescription": undefined,
      "description": "An alternative JSON file containing the configuration",
      "envVar": undefined,
      "flags": "-c, --config <MODLIST_JSON>",
      "hidden": false,
      "implied": undefined,
      "long": "--config",
      "mandatory": false,
      "negate": false,
      "optional": false,
      "parseArg": undefined,
      "presetArg": undefined,
      "required": true,
      "short": "-c",
      "variadic": false,
    },
    Option {
      "argChoices": undefined,
      "conflictsWith": [],
      "defaultValue": false,
      "defaultValueDescription": undefined,
      "description": "Suppress all output",
      "envVar": undefined,
      "flags": "-q, --quiet",
      "hidden": false,
      "implied": undefined,
      "long": "--quiet",
      "mandatory": false,
      "negate": false,
      "optional": false,
      "parseArg": undefined,
      "presetArg": undefined,
      "required": false,
      "short": "-q",
      "variadic": false,
    },
    Option {
      "argChoices": undefined,
      "conflictsWith": [],
      "defaultValue": false,
      "defaultValueDescription": undefined,
      "description": "Enable debug messages",
      "envVar": undefined,
      "flags": "-d, --debug",
      "hidden": false,
      "implied": undefined,
      "long": "--debug",
      "mandatory": false,
      "negate": false,
      "optional": false,
      "parseArg": undefined,
      "presetArg": undefined,
      "required": false,
      "short": "-d",
      "variadic": false,
    },
  ],
  "parent": null,
  "processedArgs": [],
  "rawArgs": [],
  Symbol(kCapture): false,
}
`;
